### PR TITLE
add supportings mutiple databases migrations.

### PR DESCRIPTION
**Pull request by songsd**
support migrating or rollbacking a list of databases configured in the
project.clj.

### CODE REVIEW

File: ragtime.core/src/ragtime/main.clj [hunk 1]
```diff
@@ -27,16 +27,34 @@
   (map verbose-migration ((load-var migration-fn))))
 
 (defn migrate [{:keys [database migrations]}]
-  (core/migrate-all
-   (core/connection database)
-   (resolve-migrations migrations)))
+  (letfn [(migrate_db [the_db]
```
Comment (michaelklishin) [inline ragtime.core/src/ragtime/main.clj]: Clojure uses dashes in multi-word function, namespaces, local, var names.

File: ragtime.core/src/ragtime/main.clj [hunk 2]
```diff
@@ -27,16 +27,34 @@
...
+                        (core/connection the_db)
+                        (resolve-migrations migrations))
+                      )
+          ]
+    (if (coll? database)
+      (doseq [db database]
+        (migrate_db db)
+        )
+      (migrate_db database)
+      )))
```
Comment (michaelklishin) [inline ragtime.core/src/ragtime/main.clj]: Please follow existing code style.

Comment (songsd) [inline ragtime.core/src/ragtime/main.clj]: I am making the pull request first time.
I  have changed the code style .Need making a pull request again?

**Comment by songsd** [comment]
sometimes i need to migrate many databases in the same time,so add this feature.Supporting config like this:
{:migrations ragtime.sql.files/migrations,
 :database
 ["jdbc:mysql://localhost:3306/db_shard1?useUnicode=true&characterEncoding=UTF-8&user=xxxx&password=xxx"
 "jdbc:mysql://localhost:3306/db_shard2?useUnicode=true&characterEncoding=UTF-8&user=xxxx&password=xxx"]
 }  
compatible with original: 
{:migrations ragtime.sql.files/migrations,
:database
"jdbc:mysql://localhost:3306/db_without_shard?useUnicode=true&characterEncoding=UTF-8&user=xxxx&password=xxx"
}

File: ragtime.core/src/ragtime/main.clj [hunk 3]
```diff
@@ -27,16 +27,34 @@
...
+                        (core/connection the-db)
+                        (resolve-migrations migrations))
+                      )
+          ]
+    (if (coll? database)
+      (doseq [db database]
+        (migrate-db db)
+        )
+      (migrate-db database)
+      )))
```
Comment (michaelklishin) [inline ragtime.core/src/ragtime/main.clj]: I don't think the rest of the codebase leaves trailing parentheses hanging on a separate line.

**Comment by michaelklishin** [comment]
If you need to update a pull request, simply push to the same branch, no need to submit a new PR (unless you submitted it against the wrong branch). Thank you!

**Comment by michaelklishin** [comment]
@weavejester @martintrojer what do you think about having this feature in Ragtime?

**Comment by songsd** [comment]
Done. Thank you!

**Comment by weavejester** [comment]
I don't think the `migrate` function should be changed. Clojure prefers simple functions over complex ones. Adding additional behaviour to functions based on the type is usually something to be avoided if there are other options.

Instead, I believe the `leiningen.ragtime` namespace should be changed to support a collection of configurations.

Also, when you submit a pull request, it's important that you respect coding and commenting conventions. So your git commit messages should match the others in the code base, and your code should be formatted in the standard Clojure style.

**Comment by songsd** [comment]
Thanks for advice and feeling sorry for my not respecting coding and commenting styles.
I agree "migrate" function should not be changed.I have looked "leiningen.ragtime" namespace,

(defn ragtime
  "Run ragtime.main with the options specified in the project file."
  [project command & args]
  (let [migrations (-> project :ragtime :migrations str)
        database   (-> project :ragtime :database)
        project    (add-ragtime-deps project)]
    (apply run project
           "-m" "ragtime.main"
           "-r" "ragtime.sql.database"
           "-d" database
           "-m" migrations
           command args)))

I am not so sure to change the code here ,like this:

(doseq [db database]
    (apply run project
           "-m" "ragtime.main"
           "-r" "ragtime.sql.database"
           "-d" db
           "-m" migrations
           command args)))

It seems like the command is executed many times when a command is issued .
@weavejester

**Comment by weavejester** [comment]
I would just have something to check whether or not `(:ragtime project)` is a collection or not.

```clojure
(when-let [builds (:ragtime project)]
  (doseq [build (if (coll? builds) builds [builds])]
    (apply run project ...)))
```

**Comment by songsd** [comment]
done

**Comment by weavejester** [comment]
You don't appear to have anything in your source code that loops over the databases. As far as I can tell, your code will fail to work. You also have some random indentation changes that need to be removed.

However, I'm currently in the middle of writing Ragtime 0.4.0, which significantly changes the codebase. I'll see if I can incorporate this change into the new version when I'm done.

**Comment by songsd** [comment]
thanks.Waiting for 0.4.0.

**Pull request closed without merging**